tests/load/pipeline/test_refresh_modes.py

```diff
@@ -0,0 +1,484 @@
+    # Keys contains only "resources" or is empty
+    assert list(state.keys()) == ["resources"] or not state
+    for value in state["resources"].values():
+        assert not value
+
+
+def assert_only_table_columns(cursor: DBApiCursor, expected_columns: Sequence[str]) -> None:
+    """Table has all and only the expected columns (excluding _dlt columns)"""
+    # Ignore _dlt columns
+    columns = [c[0] for c in cursor.native_cursor.description if not c[0].startswith("_")]
+    assert set(columns) == set(expected_columns)
+
+
+def column_values(cursor: DBApiCursor, column_name: str) -> List[Any]:
+    """Return all values in a column from a cursor"""
+    idx = [c[0] for c in cursor.native_cursor.description].index(column_name)
+    return [row[idx] for row in cursor.fetchall()]
+
+
+@dlt.source
+def refresh_source(first_run: bool = True, drop_sources: bool = False):
+    @dlt.resource
+    def some_data_1():
+        if first_run:
+            # Set some source and resource state
+            dlt.state()["source_key_1"] = "source_value_1"
+            resource_state("some_data_1")["run1_1"] = "value1_1"
+            resource_state("some_data_1")["run1_2"] = "value1_2"
+            yield {"id": 1, "name": "John"}
+            yield {"id": 2, "name": "Jane"}
+        else:
+            # Check state is cleared for this resource
+            assert not resource_state("some_data_1")
+            if drop_sources:
+                assert_source_state_is_wiped(dlt.state())
+            # Second dataset without name column to test tables are re-created
+            yield {"id": 3}
+            yield {"id": 4}
+
+    @dlt.resource
+    def some_data_2():
+        if first_run:
+            dlt.state()["source_key_2"] = "source_value_2"
+            resource_state("some_data_2")["run1_1"] = "value1_1"
+            resource_state("some_data_2")["run1_2"] = "value1_2"
+            yield {"id": 5, "name": "Joe"}
+            yield {"id": 6, "name": "Jill"}
+        else:
+            assert not resource_state("some_data_2")
+            if drop_sources:
+                assert_source_state_is_wiped(dlt.state())
+            yield {"id": 7}
+            yield {"id": 8}
+
+    @dlt.resource
+    def some_data_3():
+        if first_run:
+            dlt.state()["source_key_3"] = "source_value_3"
+            resource_state("some_data_3")["run1_1"] = "value1_1"
+            yield {"id": 9, "name": "Jack"}
+            yield {"id": 10, "name": "Jill"}
+        else:
+            assert not resource_state("some_data_3")
+            if drop_sources:
+                assert_source_state_is_wiped(dlt.state())
+            yield {"id": 11}
+            yield {"id": 12}
+
+    @dlt.resource
+    def some_data_4():
+        yield []
+
+    yield some_data_1
+    yield some_data_2
+    yield some_data_3
+    yield some_data_4
+
+
+@pytest.mark.parametrize(
+    "destination_config",
+    destinations_configs(default_sql_configs=True, subset=["duckdb"]),
```

Comment:
why only duckdb?

Reply:
as mentioned below, you should be able to test the filesystems with the same test

Reply:
Only variance could be the implementation of "drop/truncate tables" methods, but they're already tested elsewhere. So didn't think it would be necessary to run other destinations here. Just going by the assumption that all sql destinations have their interface tested.    
But maybe doesn't hurt to be thorough if something changes with the implementation in the future.